File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-3.cpp

class Solution {
public:
    void dfs(int r , int c , vector<vector<bool>>& visited , int rows, int cols ,vector<vector<int>>& heights){
        vector<vector<int>> dirs{{1,0},{-1,0},{0,1},{0,-1}};
        visited[r][c]=true ;
        for(auto& d : dirs){
            int i= r+d[0];
            int j = c+d[1];
            if(i>=0 && j>=0 && i<rows && j<cols && !visited[i][j] && heights[i][j]>=heights[r][c])
                  dfs(i,j,visited,rows,cols,heights);
        }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int rows = heights.size();
        int cols = heights[0].size();
        vector<vector<bool>> pacific(rows, vector<bool>(cols, false));
        vector<vector<bool>> atlantic(rows, vector<bool>(cols, false));
        for(int i=0;i<rows;i++){
            dfs(i,0,pacific,rows ,cols,heights);
            dfs(i,cols-1,atlantic,rows,cols,heights);
        }
        for(int j=0;j<cols;j++){
            dfs(0,j,pacific,rows,cols,heights);
            dfs(rows-1,j,atlantic,rows,cols,heights);
        }
        vector<vector<int>> res ;
        for(int i=0;i<rows;i++)
           for(int j=0;j<cols;j++) 
             if(pacific[i][j] && atlantic[i][j])
                res.push_back({i,j});

        return res;
    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-3.java
class Solution {
    public void capture(int i, int j , char[][] board){
        if(i<0 || j<0 || i>= board.length || j>= board[0].length || board[i][j]!='O')
           return ;
        board[i][j]='T';
        capture(i+1,j,board);
        capture(i-1,j,board);
        capture(i,j+1,board);
        capture(i,j-1,board);
    }
    public void solve(char[][] board) {
        int rows = board.length;
        int cols = board[0].length ;
        for(int i=0; i<rows ;i++){
            capture(i,0,board);
            capture(i,cols-1,board);
        }
        for(int j=0; j<cols ;j++){
            capture(0,j,board);
            capture(rows-1,j,board);
        }

        for(int i=0;i<rows;i++)
            for(int j=0;j<cols;j++){
               if(board[i][j] == 'T')
                    board[i][j]='O';
                else if (board[i][j]=='O')
                    board[i][j]='X';
            }

    }
}
